systems/controllers/QPCommon.cpp

```diff
@@ -381,7 +399,21 @@ int setupAndSolveQP(NewQPControllerData *pdata, std::shared_ptr<drake::lcmt_qp_c
     num_active_contact_pts += iter->contact_pts.size();
   }
 
-  pdata->r->HandC(robot_state.q,robot_state.qd,(MatrixXd*)nullptr,pdata->H,pdata->C,(MatrixXd*)nullptr,(MatrixXd*)nullptr,(MatrixXd*)nullptr);
+  // handle external wrenches to compensate for
+  if (qp_input->body_wrench_data.size() > 0) {
+    const int wrench_size = 6;
+    MatrixXd f_ext(wrench_size, pdata->r->num_bodies);
+    f_ext.setZero();
+    for (auto it = qp_input->body_wrench_data.begin(); it != qp_input->body_wrench_data.end(); ++it) {
+      const drake::lcmt_body_wrench_data& body_wrench_data = *it;
+      Map< const Matrix<double, wrench_size, 1> > wrench(body_wrench_data.wrench);
+      f_ext.col(body_wrench_data.body_id - 1) = wrench;
+    }
+    pdata->r->HandC(robot_state.q,robot_state.qd,&f_ext,pdata->H,pdata->C,(MatrixXd*)nullptr,(MatrixXd*)nullptr,(MatrixXd*)nullptr);
```

Comment:
This `wrench` is the wrench applied at the body origin, in the body frame, is it correct?